tweak/balance: логика работы радиации и радия на нуклеациях

shahribek opened the pull request:
<!-- Если вам необходимо отключить IconDiffBot2 или MapDiffBot2, тогда добавьте в название вашего пр-а тэги [IDB IGNORE] или [MDB IGNORE] соответственно. -->

<!-- Вы можете совмещать до 2-х тэгов, например, balance/tweak:
Список возможных тэгов для пр-а:
- add: Вы добавляете что-то новое.
- admin: Вы меняете что-то связанное с администрацией. (кнопки, управление, панели, щитспавн и т.д.)
- balance: Вы производите балансировку в игре.
- bugfix: Вы исправили некий баг.
- code_imp: Вы имплементируете новое для билда, не меняя при этом ничего в самой игре.
- config: Вы меняете конфиг или работу SQL.
- del: Вы удаляете что-то из игры.
- experiment: Ваш ПР сделан с целью какого-то эксперимента.
- map: Вы меняете только карту.
- image: Вы добавляете/удаляете спрайты.
- sound: Вы добавляете/удаляете звуки.
- spellcheck: Вы исправили опечатку.
- local: Вы добавляете новый текст на русском или переводите на русский.
- tweak: Вы внесли незначительную правку.
- refactor: Вы полностью переписали старый код, улучшив его, НО не изменив функционал.
- server: Вы меняете что-то связанное с серверной частью или Github.
- wip: Ваш PR в драфте и планируется длительная разработка. -->

## Что этот ПР делает
убирает весь код который опционален и убирает коды которые никогда не работали и никогда не будут.
дополнительно меняет логику работы радиации и радия. полностью удаляя кастомную проверку на наличие радия на проверку есть ли в теле радиация.
P.S. это рвеменное решение пока я полностью не доработаю всю рассу до соотвествии с лором
<!-- Опишите, что делает ваш пулл-реквест, укажите основные изменения. -->
<!-- В случае исправления бага, укажите ссылку на канал в #баг-репорты-v2. В ином случае, опишите баг и его воспроизведение. -->

## Почему это хорошо для игры
круто, стабильно(наверное), дополнительно обьясняет почему радий лечит нуклеатов
<!-- Опишите, почему, по вашему, следует добавить эти изменения в билд. -->


## Тестирование
дебаггинг провёл. Всё работает довольно стабильно.
<!-- Как Вы тестировали свои изменения? Ведь тестировали? -->

shahribek committed:
added and applied newtrait + removed radimmunity from nucleates

shahribek committed:
added and used new trait for avoiding radiation effects

shahribek committed:
removed custom check for radium for nucleations

shahribek committed:
removed codes which has never worked and never will

shahribek committed:
Merge branch 'ss220-space:master220' into nuclie-radiation-hell

shahribek committed:
апечатка

shahribek commented:
так. проверка проведена. пора решать проблемы

shahribek committed:
Trait validity check fix

shahribek committed:
: Alphabetical ordering violation fix

shahribek committed:
another one...

vovakons commented on code/modules/mob/living/carbon/human/life.dm:
```diff
@@ -184,19 +184,8 @@
-						return
-
-		if(!HAS_TRAIT(src, TRAIT_RADIMMUNE))
+
```
лишняя пустая строка

vovakons commented on code/modules/mob/living/carbon/human/life.dm:
```diff
@@ -249,6 +238,32 @@
+				radiation = max(radiation - heal_cost, 0)
+				heal_damages(heal_brute, heal_burn)
+
+			radiation = max(radiation-1, 0)
```
вынести бы все внутри этого if в отдельный прок типа `process_radiation_healing`

vovakons replied:
`radiation = max(radiation - 1, 0)`

shahribek replied:
тут же буквально сам процесс называется `handle_mutations_and_radiation()`
не сойдёт?

vovakons commented on code/modules/mob/living/carbon/human/life.dm:
```diff
@@ -249,6 +238,32 @@
+
+			else if(radiation >= heal_cost)
+				radiation = max(radiation - heal_cost, 0)
+				heal_damages(heal_brute, heal_burn)
```
```suggestion
			heal_mod = 1
			if(health < HEALTH_THRESHOLD_CRIT && radiation >= heal_cost * 2)
				heal_mod = 2
			if(radiation >= heal_cost * heal_mod)
				radiation = max(radiation - heal_cost * heal_mod, 0)
				heal_damages(heal_brute * heal_mod, heal_burn * heal_mod)
```

shahribek replied:
`var/heal_mod = 1`*

shahribek replied:
так она будет ждать накопление двойной порции. имеющийся не ждёт такого и использует обычную порцию если на двойную не хватает.

shahribek replied:
а хотя.
кажется я устал, мозги отказываются работать

shahribek committed:
Update code/modules/mob/living/carbon/human/life.dm

обновлено по предложению

Co-authored-by: VladimirKonstantinov <anbu23477@gmail.com>

shahribek committed:
вёрб фикс

shahribek committed:
excess line removal and extra space addition

shahribek committed:
ошибся местом

shahribek committed:
вывел весь процесс в отдельный прок

shahribek committed:
proc defined

shahribek committed:
remembered why I did that

shahribek committed:
emembered something really important

shahribek committed:
Revert "emembered something really important"

This reverts commit 9c56c0cc42035274766aeca2bfebf5f99339bf4d.

shahribek committed:
Revert "remembered why I did that"

This reverts commit 7c8c8af1bf3271b325d815fb6aa8d3a6d8cf28df.

shahribek committed:
made an additional option for get-damaged_organs()

remote-shah committed:
bugfix

shahribek committed:
used new option in proc

shahribek committed:
made new proc for identifying broken limbs

shahribek committed:
low priority limb factures healing state

shahribek committed:
quick fix

shahribek committed:
quick fix 2

shahribek committed:
quick fix 3

shahribek committed:
quick fix 4
*sigh*

shahribek committed:
quick fix 5
*cough*

shahribek commented on code/modules/mob/living/carbon/human/human_damage.dm:
```diff
@@ -301,14 +301,32 @@
+				if(!(type_flags & AFFECT_ORGANIC_ORGAN) && !int_organ.is_robotic())
+					continue
+				parts += int_organ
+	return parts
```
так. Не понял, почему тут выдаётся ошибка что игра считает `return parts` за `returnparts`

shahribek replied:
починил. неважно

shahribek committed:
quick fix 6
"I want to kill copilot. Totally useless machine"

shahribek committed:
quick fix 7

shahribek committed:
quick fix 8
*sigh* finally

shahribek committed:
used existing proc instead of custom

shahribek committed:
runtime error fix

shahribek committed:
updated internal and fractures healing cost

shahribek committed:
Merge branch 'master220' into nuclie-radiation-hell

shahribek committed:
опечатка

shahribek committed:
a small nerf after debug

shahribek committed:
Merge branch 'master220' into nuclie-radiation-hell

shahribek committed:
possible bug fix

shahribek committed:
simpler logic

shahribek committed:
second step

shahribek committed:
Revert "second step"

This reverts commit af1b4d76d71cd93261115b30f1ffc087a08084ed.

shahribek commented:
забыл ветки поменять

shahribek committed:
Merge branch 'master220' into nuclie-radiation-hell

dageavtobusnick commented on code/modules/mob/living/carbon/human/human_damage.dm:
```diff
@@ -300,16 +300,27 @@
+			if(!(flags & AFFECT_ORGANIC_ORGAN) && !int_organ.is_robotic())
 				continue
-			parts += bodypart
+			parts += int_organ
```
Обопщи, выделив общий код в отдельный прок

shahribek committed:
split proc functions

shahribek committed:
slight forgotten change

shahribek committed:
variable name logic

shahribek committed:
quick fix

shahribek committed:
Merge branch 'master220' into nuclie-radiation-hell

shahribek committed:
optimization

dageavtobusnick commented on code/modules/mob/living/carbon/human/human_damage.dm:
```diff
@@ -312,6 +312,18 @@
+			if(!(flags & AFFECT_ORGANIC_ORGAN) && !int_organ.is_robotic())
+				continue
+			organs += int_organ
+	return organs
```
Ты не понял. У тебя буквально ддя интерналов и для экстерналов делаются одинаковые проверки. Проки буквально копируют друг друга. Можно сделать прок, в который будешь передавать список органов(внешний или внутрений) и итерировать внутри как obj/item/organ

shahribek replied:
кажется понял, скоро будет готово

shahribek committed:
proc tweak
this was a pain.

shahribek committed:
Merge branch 'master220' into nuclie-radiation-hell

shahribek committed:
Merge branch 'master220' into nuclie-radiation-hell

Antoonij reviewed (commented):
лучше будет сделать весь этот механ лечения компонентом и выдавать тем, кому нужно. Это избавит от лишних трейтов и позволит твикать циферки лечения за кол-во радиации и т.д

dageavtobusnick commented:
> лучше будет сделать весь этот механ лечения компонентом и выдавать тем, кому нужно. Это избавит от лишних трейтов и позволит твикать циферки лечения за кол-во радиации и т.д

По идее даже элемента хватит.

shahribek commented:
> лучше будет сделать весь этот механ лечения компонентом и выдавать тем, кому нужно. Это избавит от лишних трейтов и позволит твикать циферки лечения за кол-во радиации и т.д

могу и так сделать. Раз нужно меньше трейтов

shahribek committed:
Merge branch 'master220' into nuclie-radiation-hell

shahribek committed:
Merge branch 'master220' into nuclie-radiation-hell

shahribek committed:
multiple changes
stopped using traits
made radimmunity to ignore all effects instead of blocking rad gain

shahribek committed:
quick fix

dageavtobusnick commented:
Ты совсем не понял, что от тебя хотели. Посмотри, что такое компоненты, элементы и сигналы и переделай на них. Проблема не в самих лишних трейтах

shahribek commented:
> Ты совсем не понял, что от тебя хотели. Посмотри, что такое компоненты, элементы и сигналы и переделай на них. Проблема не в самих лишних трейтах

извиняюсь. Слишком заджунил

shahribek commented:
бля вместо отмены промазал

shahribek committed:
Merge branch 'master220' into nuclie-radiation-hell

dageavtobusnick commented:
Из драфта выводи, когда переделаешь на элемент или компонент.

shahribek committed:
Merge branch 'master220' into nuclie-radiation-hell

shahribek committed:
Revert "quick fix"

This reverts commit d0bbcad1281b65f0bc9135d6d90bb62d602e58e9.

shahribek committed:
Revert "multiple changes"

This reverts commit 8d7ae7b1e10a630ea8b58792c593210e5bca5bed.

shahribek committed:
Merge branch 'master220' into nuclie-radiation-hell

shahribek committed:
elementized

shahribek committed:
including new file

shahribek committed:
quick fix

shahribek committed:
usage of elements
ugh;
just need to test now.

shahribek committed:
deleted the radiation_healing trait

shahribek commented:
вроде кое как разобрался. Осталось только протестировать

shahribek committed:
Merge branch 'master220' into nuclie-radiation-hell

shahribek committed:
Merge branch 'master220' into nuclie-radiation-hell

shahribek committed:
Merge branch 'master220' into nuclie-radiation-hell

shahribek committed:
исправления множество комментариев для большей ясности
добавление новой функции для get_damaged_organs() для ввода минимальной требуемого урона для внутренних органов

shahribek committed:
small fix

shahribek committed:
code cleanup

shahribek committed:
Merge branch 'master220' into nuclie-radiation-hell

shahribek commented:
/gemini review

shahribek commented on code/datums/elements/radiation_healing.dm:
```diff
@@ -0,0 +1,74 @@
+			human.radiation = max(human.radiation - FRACTURE_HEAL_COST, 0)
+
+	if(HAS_TRAIT(human, TRAIT_NO_RADIATION_EFFECTS)) // passive radiation drain in case if it doesn't drain itself
+		human.radiation = max(human.radiation - PASSIVE_RAD_DRAIN, 0)
```
всё так и задумано, но при использовании реагента "радиум" из-за малого количество радиации одно будет работать первее чем другое

shahribek committed:
more randomness
*sigh*

shahribek commented on code/datums/elements/radiation_healing.dm:
```diff
@@ -0,0 +1,74 @@
+		human.heal_damages(heal_brute * heal_mod, heal_burn * heal_mod)
+
+	//internal healing in medium priority. Using radium won't heal internal damage while healing external damage
+	var/list/obj/item/organ/internal/int_damaged_organs = human.get_damaged_organs(flags = AFFECT_ORGANIC_INTERNAL_PARTS)
```
@gemini-code-assist разве мой вариант уже не говорит буквально что она возвращяет список повреждённых органов?

shahribek committed:
defined nucleation light color

shahribek committed:
Merge branch 'master220' into nuclie-radiation-hell

shahribek committed:
bug fix

shahribek committed:
bug fix

shahribek committed:
linters Grep check error fix

shahribek committed:
proc logic tweak & bugfix

shahribek commented:
/gemini review

shahribek committed:
maximum radiation definition

shahribek committed:
comment rewrite

shahribek committed:
small tweak

shahribek committed:
bug_fix

shahribek commented:
/gemini review

shahribek commented on code/datums/elements/radiation_healing.dm:
```diff
@@ -0,0 +1,74 @@
+		var/obj/item/organ/external/limb = pick(fractured_limbs)
+		if(limb.has_fracture())//double check as it might have been healed already
+			limb.mend_fracture()
+			human.radiation = max(human.radiation - FRACTURE_HEAL_COST, 0)
```
возможно стоит подумать над модульностью элемента чем над комментами

shahribek replied:
@gemini-code-assist может всё это дело сделать более модульным чем менять то что так и задумано что бы и другие люди могли её настроить под себя задавая только нужные аргументы?

shahribek committed:
Merge branch 'master220' into nuclie-radiation-hell

shahribek committed:
Merge branch 'master220' into nuclie-radiation-hell

dageavtobusnick commented on code/modules/mob/living/carbon/human/human_damage.dm:
```diff
@@ -301,17 +301,36 @@
+	if(AFFECT_INTERNAL_ORGANS & flags)
+		organs += internal_organs
+	for(var/obj/item/organ/organ as anything in organs)
+		var/external = istype(organ, /obj/item/organ/external)
```
используй изхелпер. Если его еще нету - сделай.

dageavtobusnick commented on code/modules/mob/living/carbon/human/human_damage.dm:
```diff
@@ -301,17 +301,36 @@
+			damaged = (brute && bodypart.brute_dam) || (burn && bodypart.burn_dam)
+		else
+			var/obj/item/organ/internal/int_organ = organ
+			damaged = int_organ.damage > 0
```
Это лучше сделать через прок(если его еще нету) и переопределить его для экстерналов и интерналов, а не городить этот велосипед.

dageavtobusnick commented on code/datums/elements/radiation_healing.dm:
```diff
@@ -0,0 +1,74 @@
+
+	if(human.radiation >= FRACTURE_HEAL_COST && length(fractured_limbs))
+		var/obj/item/organ/external/limb = pick(fractured_limbs)
+		if(limb.has_fracture())//double check as it might have been healed already
```
Чем она может быть уже вылечена? у нас нет многопоточности в самом бьенде, а выполнение прока может прерваться только через sleep и проки, что его используют. К чему тут перепроверка?

shahribek replied:
я сам на знаю, просто подумал что такой исход возможен.

shahribek committed:
small tweak

shahribek committed:
Merge branch 'master220' of https://github.com/shahribek/Paradise into nuclie-radiation-hell

shahribek committed:
tweak

shahribek committed:
Merge branch 'master220' into nuclie-radiation-hell

ROdenFL commented on code/datums/elements/radiation_healing.dm:
```diff
@@ -0,0 +1,97 @@
+
+	var/rads = human.radiation
+	if(rads <= 0)
+		return
```
```suggestion
	if(human.radiation <= 0)
		return
```

ROdenFL commented on code/datums/elements/radiation_healing.dm:
```diff
@@ -0,0 +1,97 @@
+		heal_mod = 2
+
+	if(human.radiation >= heal_cost * heal_mod)
+		human.radiation = max(human.radiation - heal_cost * heal_mod, 0)
```
Вроде и так уже уверены, что радиация больше
```suggestion
		human.radiation = human.radiation - heal_cost * heal_mod
```

ROdenFL commented on code/datums/elements/radiation_healing.dm:
```diff
@@ -0,0 +1,97 @@
+
+	var/heal_mod = 1
+
+	if(human.isInCrit() && human.radiation >= heal_cost * 2)
+		heal_mod = 2
+
+	if(human.radiation >= heal_cost * heal_mod)
```
Смущает моментик, что в Крите наше лечение должно усиливаться, но при этом из-за увеличения модификатора, лечение которое бы прошло по старой цене, по новой не пройдет. При радиации 3 мы можем позволить цену лечения 2, но впав в Крит не вывезем цену 4 и не будем лечиться совсем. Странная механика.

shahribek replied:
можно накопить до 800 радиации в теле чего хватит ровно на 800 тиков и то что от суперматерии тоже можно кучу радиации словить, так что в таких случаях выгоднее будет за двойную цену чем всё так же медленно

ROdenFL replied:
я про случай когда на двойную цену не хватает, а на обычную хватает

shahribek replied:
тогда будет обычный хил как и всегда. На то и стоит проверка "хватает ли радиации с модификатором?" в строке 61

ROdenFL commented on code/modules/mob/living/carbon/human/human_damage.dm:
```diff
@@ -301,17 +301,29 @@
+	// returns a list of damaged organs that meet the given criteria
+	// brute and burn are boolean values for checking that type of damage
+	// flags determine what kind of organs to include, by default is set to AFFECT_ALL_EXTERNAL_PARTS
+	// see more in code/__DEFINES/flags.dm line 179-189
```
А ваш комментарий и тут и там показывают....

shahribek replied:
я не лажу с комментариями

shahribek committed:
Update code/datums/elements/radiation_healing.dm

ROdenFL's suggestion

Co-authored-by: ROdenFL <ROdenFL@yandex.ru>

shahribek committed:
Merge branch 'master220' into nuclie-radiation-hell

shahribek committed:
Update code/datums/elements/radiation_healing.dm

another suggestion

Co-authored-by: ROdenFL <ROdenFL@yandex.ru>

shahribek committed:
optimization

shahribek committed:
coment mistake

ROdenFL commented on code/modules/mob/living/carbon/human/human_damage.dm:
коменты обычно над проком

shahribek replied:
окей

shahribek committed:
another comment change

shahribek committed:
Merge branch 'master220' into nuclie-radiation-hell

shahribek committed:
Merge branch 'master220' into nuclie-radiation-hell
